internal/email/email.go

```diff
@@ -61,56 +108,84 @@ func NewMessage(
-		return getEmailBodyText(inviteURL, sponsor, project, role)
+		zerolog.Ctx(ctx).Error().Err(err).Msg("error creating the HTML template for email invitations")
+		return getEmailBodyText(ctx, data)
 	}
-	// Execute the template
-	var b strings.Builder
-	if err := tmpl.Execute(&b, data); err != nil {
-		return ""
+	err = bodyHTMLTmpl.Execute(ctx, &b, data, EmailBodyMaxLength)
+	if err != nil {
+		zerolog.Ctx(ctx).Error().Err(err).Msg("error executing the HTML template for email invitations")
+		return getEmailBodyText(ctx, data)
 	}
-	return b.String()
+	return b.String(), nil
 }
 
-// getEmailBodyText returns the text body for the email based on the message payload
-func getEmailBodyText(inviteURL, sponsor, project, role string) string {
-	return fmt.Sprintf("You have been invited to join %s as %s by %s. Visit %s to accept the invitation.",
-		project, role, sponsor, inviteURL)
+// getTextBodyString returns the text body for the email based on the message payload
+func getEmailBodyText(ctx context.Context, data bodyData) (string, error) {
+	var b strings.Builder
+	bText := bodyText
+
+	bodyTextTmpl, err := util.NewSafeHTMLTemplate(&bText, "body-invite-text")
+	if err != nil {
+		zerolog.Ctx(ctx).Error().Err(err).Msg("error creating the text template for email invitations")
+		return "", err
+	}
+	err = bodyTextTmpl.Execute(ctx, &b, data, EmailBodyMaxLength)
+	if err != nil {
+		zerolog.Ctx(ctx).Error().Err(err).Msg("error executing the text template for email invitations")
+		return "", err
+	}
+	return b.String(), nil
 }
 
 // getEmailSubject returns the subject for the email based on the message payload
 func getEmailSubject(project string) string {
 	return fmt.Sprintf("You have been invited to join the %s organization in Minder", project)
 }
+
+// isValidField checks if the string is empty or contains HTML or JavaScript injection
+func isValidField(str string) error {
+	// Check for HTML content
+	escapedHTML := template.HTMLEscapeString(str)
+	if escapedHTML != str {
+		return fmt.Errorf("string %s contains HTML injection", str)
+	}
+
+	// Check for JavaScript content separately
+	escapedJS := template.JSEscaper(str)
+	if escapedJS != str {
+		return fmt.Errorf("string %s contains JavaScript injection", str)
+	}
+
+	return nil
+}
+
+// validateDataSourceTemplate validates the template data source for HTML injection attacks
+func validateDataSourceTemplate(s interface{}) error {
+	// Get the reflect.Value of the pointer to the struct
+	v := reflect.ValueOf(s).Elem()
+
+	// Iterate over the fields of the struct
+	for i := 0; i < v.NumField(); i++ {
+		field := v.Field(i)
+
+		// Check if the field is settable and of kind string
+		if field.CanSet() && field.Kind() == reflect.String {
+			strVal := field.String()
+
+			// Execute your function on the field value
+			err := isValidField(strVal)
+			if err != nil {
+				return fmt.Errorf("field %s failed validation - %s", v.Type().Field(i).Name, strVal)
+			}
+		}
+	}
+	return nil
```

Comment:
I'm not thrilled with the reflection here -- in particular, we should know that each field is settable and of type String, so we should be able to error if it is not.  Similarly, I'd prefer to have a more concrete argument than `any`, because this doesn't sanitize other fields, like if we added something which contained further string fields, or implements `fmt.Stringer()`, but is not a string itself.

Reply:
Thanks 👍 I updated the reflect part to not take any as input but also to enforce that all fields are of type string. If this happens to be wrong, we'll fail the validation and error out.